prow/cmd/exporter/collector.go

```diff
@@ -169,3 +171,153 @@ func sanitizeLabelName(s string) string {
 func escapeString(v string) string {
 	return escapeWithDoubleQuote.Replace(v)
 }
+
+type prowJobLifecycleCollector struct {
+	lastKnownState map[types.UID]lastState
+	HistogramVec   *prometheus.HistogramVec
+}
+
+type lastState struct {
+	state    prowapi.ProwJobState
+	observed time.Time
+}
+
+func (pjc *prowJobLifecycleCollector) Update(oldJob *prowapi.ProwJob, newJob *prowapi.ProwJob) {
+	now := time.Now()
+	if lastKnownState, exists := pjc.lastKnownState[newJob.UID]; !exists {
+		if newJob.Status.State == prowapi.TriggeredState {
+			now = newJob.CreationTimestamp.Time
+			pjc.lastKnownState[newJob.UID] = lastState{
+				state:    newJob.Status.State,
+				observed: now,
+			}
+			return
+		} else if oldJob == nil || oldJob.Status.State == newJob.Status.State {
+			// we can't determine the time when the job entered the current state, do nothing
+			// TODO: we can reconstruct some states by looking at the pods
+			return
+		}
+		// first time we see a state transition for a job live, or the job is still pending, now we record the time
```

Comment:
should have been "or the job is `triggered`" instead of `pending`.

Reply:
Triggered is not even a phase today, we should make it one

Reply:
Hm, as far as I know it is. Every job which I see created is initially in `triggered`.